[Issue 88] Better documentation for the MSI credentials, and support for the IDENTITY_* environment variables (#97)
Azure/ms-rest-nodeauth: README.md

# ms-rest-nodeauth [![Build Status](https://dev.azure.com/azure-public/adx/_apis/build/status/public.Azure.ms-rest-nodeauth)](https://dev.azure.com/azure-public/adx/_build/latest?definitionId=9)

This library provides different node.js based authentication mechanisms for services in Azure. It also contains rich type definitions thereby providing a good TypeScript experience.

All the authentication methods support callbacks as well as promises. If they are called within an async method in your application then you can use the async/await pattern as well.

Things to consider:

- Many of these authentication methods accept a `domain` to be specified, which accepts `tenants` to be passed in.
- For personal accounts, credentials created without passing a domain won't be able to access most of the account resources automatically.
- In this case you will need to call `buildTenantList` to gather the list of tenants so that the ID of one of them can be passed into the `setDomain` method of the returned credential.
  Once the domain is set, you will be able to access resources from subscriptions in that tenant.

> 

Reviewer: The above is for the change in #91. Assuming that this PR is mainly about MSI, we should remove it

```

### MSI (Managed Service Identity) based login from a virtual machine created in Azure.

> The code below works for both system managed and user-assigned managed identities. You can leave the `options` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId`. If your VM has multiple user-assigned managed identities, you must include `objectId` and `identityId`.

Reviewer: Sorry about nitpicking. When reading this there is a chance that one might think that clientId, objectId are things that need to be passed in directly. So adding options like the below should help.

```suggestion
The code below works for both system managed and user-assigned managed identities. You can leave the `options` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId` in the options. If your VM has multiple user-assigned managed identities, you must include `objectId` and `identityId` in the options as well.
```

> The code below works for both system managed and user-assigned managed identities. You can leave the `MSIVmOptions` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId`. If your VM has multiple user-assigned managed identities, you must include `objectId` and `identityId`.

Reviewer: We should have a similar write up for the App Service creds in the next section.

> The code below works for both system managed and user-assigned managed identities. You can leave the `MSIVmOptions` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId`. If your VM has multiple user-assigned managed identities, you must include `objectId` and `identityId`.

Reviewer: Since `MSIVmOptions` is only an interface, should we refer to it as "options" instead as that is the variable that should be kept empty or updated in the code snippet? Same feed back for the app service code snippet

```suggestion
The code below works for both system managed and user-assigned managed identities. You can leave the `options` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId`. If your VM has multiple user-assigned managed identities, you must include `objectId` and `identityId`.
```

```typescript
import * as msRestNodeAuth from "@azure/ms-rest-nodeauth";

const options: msRestNodeAuth.MSIVmOptions = {
  // The objectId of the managed identity you would like the token for.
  // Required, if your VM has multiple user-assigned managed identities.
  //
  //     objectId: "your-managed-identity-object-id",
  //

  // The clientId of the managed identity you would like the token for.

>   // Required, if your VM has multiple user-assigned managed identities.

Reviewer: Based on the write up above, we need client id even for single user-assigned managed identity right?

```suggestion
  // Required, if your VM has any user-assigned managed identity.
```

```

### MSI (Managed Service Identity) based login from a virtual machine created in Azure.

> The code below works for both system managed and user managed identities. You can leave the `MSIVmOptions` empty if you want to use system managed identity. If you want to use the user managed identity, you must at least provide the `clientId`.

Reviewer: Is there any difference between the case of single user managed identity vs multiple user-assigned managed identities? Because here we talk about "user managed identity", but in the sample code below we talk about "multiple user-assigned managed identities"

Author: Yes there is. I'll update this text.

```

### MSI (Managed Service Identity) based login from an AppService or Azure Function created in Azure.

The code below works for both system managed and user-assigned managed identities. You can leave the `MSIAppServiceOptions` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId`.

```typescript
import * as msRestNodeAuth from "@azure/ms-rest-nodeauth";

const options: msRestNodeAuth.MSIAppServiceOptions = {

  // The clientId of the managed identity you would like the token for.
  // Required, if your app service has user-assigned managed identities.

>   clientId: "your-managed-identity-client-id"

Reviewer: This part should be commented just like in the previous code snippet. Users can uncomment and add their value if they are using user-assigned managed identity

```

### MSI (Managed Service Identity) based login from an AppService or Azure Function created in Azure.

> The code below works for both system managed and user-assigned managed identities. You can leave the `MSIAppServiceOptions` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId`.

Reviewer: ```suggestion
The code below works for both system managed and user-assigned managed identities. You can leave the `options` empty if you want to use system managed identity. If you want to use the user-assigned managed identity, you must at least provide the `clientId` in the options.
```

import * as msRestNodeAuth from "@azure/ms-rest-nodeauth";

const options: msRestNodeAuth.MSIAppServiceOptions = {
  // The local URL from which your app can request tokens.
  // For example: `export MSI_ENDPOINT="http://127.0.0.1:41741/MSI/token/"`.
  // Many Azure services have the environment variables "MSI_ENDPOINT" or "IDENTITY_ENDPOINT" set.
  // Our client will try to load the endpoint from any of these environment variables by default.
  //
  //     msiEndpoint: process.env["MSI_ENDPOINT"] || process.env["IDENTITY_ENDPOINT"] || "http://127.0.0.1:41741/MSI/token/",
  //

  // The clientId of the managed identity you would like the token for.
  // Required, if your app service has user-assigned managed identities.

>   clientId: "48f97062-a6f3-48ae-b05b-e6df3468c256"

Reviewer: In all such cases where user has to add a value, we should use something like `"your-managed-identity-client-id"`
